File: design-pattern-jvm/app/src/main/java/design/pattern/java/patterns/behavioral/observer/NewsPaper.java
package design.pattern.java.patterns.behavioral.observer;

import java.util.ArrayList;
import java.util.List;

public class NewsPaper {
  List<NewsPaperReader> readers;

  public NewsPaper() {
    this.readers = new ArrayList<>();
  }

  public NewsPaperSubscription subscribe(NewsPaperReader reader) {
    this.readers.add(reader);
    return new NewsPaperSubscription(this, reader);
  }

  public void notify(Notice news) {
    this.readers.forEach(reader -> reader.apply(news));
  }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/behavioral/observer/NewsPaper.kt
package design.pattern.kotlin.patterns.behavioral.observer

class NewsPaper() {
  var readers: MutableList<NewsPaperReader> = mutableListOf()

  fun subscribe(reader: NewsPaperReader): NewsPaperSubscriptions {
    this.readers.add(reader)
    return NewsPaperSubscriptions(this, reader)
  }

  fun notify(news: Notice) {
    this.readers.forEach { it.invoke(news) }
  }
}
